dataplane/handlers/interface.go

```diff
@@ -0,0 +1,293 @@
+		log.Info("reconciling interfaces")
+		root, ok := val.Val()
+		if !ok || root.Interface == nil {
+			return ygnmi.Continue
+		}
+		for _, i := range root.Interface {
+			ni.reconcile(i)
+		}
+		return ygnmi.Continue
+	})
+
+	ni.linkDoneCh = make(chan struct{})
+	linkUpdateCh := make(chan netlink.LinkUpdate)
+	ni.addrDoneCh = make(chan struct{})
+	addrUpdateCh := make(chan netlink.AddrUpdate)
+
+	if err := netlink.LinkSubscribe(linkUpdateCh, ni.linkDoneCh); err != nil {
+		return fmt.Errorf("failed to sub to link: %v", err)
+	}
+	if err := netlink.AddrSubscribe(addrUpdateCh, ni.addrDoneCh); err != nil {
+		return fmt.Errorf("failed to sub to addr: %v", err)
+	}
+
+	go func() {
+		for {
+			select {
+			case up := <-linkUpdateCh:
+				ni.handleLinkUpdate(&up)
+			case up := <-addrUpdateCh:
+				ni.handleAddrUpdate(&up)
+			}
+		}
+	}()
+
+	go func() {
+		// TODO: handle error
+		if _, err := watcher.Await(); err != nil {
+			log.Warningf("interface watch err: %v", err)
+		}
+	}()
+
+	return nil
+}
+
+// Stop stops all watchers.
+func (ni *Interface) Stop() {
+	ni.watchCancelFn()
+	close(ni.linkDoneCh)
+	close(ni.addrDoneCh)
+}
+
+// reconcile compares the interface config with state and modifies state to match config.
+func (ni *Interface) reconcile(config *oc.Interface) {
+	ni.mu.RLock()
+	defer ni.mu.RUnlock()
+
+	tapName := engine.IntfNameToTapName(config.GetName())
+	state := ni.state[config.GetName()]
+
+	if config.GetOrCreateEthernet().MacAddress != nil {
+		if config.GetEthernet().GetMacAddress() != state.GetEthernet().GetMacAddress() {
+			log.Infof("setting interface %s hw-addr %q", engine.IntfNameToTapName(config.GetName()), config.GetEthernet().GetMacAddress())
+			if err := kernel.SetInterfaceHWAddr(engine.IntfNameToTapName(config.GetName()), config.GetEthernet().GetMacAddress()); err != nil {
+				log.Warningf("Failed to set mac address of port: %v", err)
+				return
+			}
+		}
+	}
+	if config.GetOrCreateSubinterface(0).Enabled != nil {
+		if state.GetSubinterface(0).Enabled == nil || config.GetSubinterface(0).GetEnabled() != state.GetSubinterface(0).GetEnabled() {
+			log.Infof("setting interface %s enabled %t", engine.IntfNameToTapName(config.GetName()), config.GetSubinterface(0).GetEnabled())
+			if err := kernel.SetInterfaceState(engine.IntfNameToTapName(config.GetName()), config.GetSubinterface(0).GetEnabled()); err != nil {
+				log.Warningf("Failed to set state address of port: %v", err)
+				return
+			}
+		}
+	}
+	// TODO: refactor this.
+	for _, addr := range config.GetOrCreateSubinterface(0).GetOrCreateIpv4().Address {
+		configIP := addr.Ip
+		configPL := addr.PrefixLength
+		var stateIP *string
+		var statePL *uint8
+		stateAddr := state.GetSubinterface(0).GetIpv4().GetAddress(addr.GetIp())
+		if stateAddr != nil {
+			stateIP = stateAddr.Ip
+			statePL = stateAddr.PrefixLength
+		}
+
+		if configIP != nil && configPL != nil && (stateIP == nil || *statePL != *configPL) {
```

Comment:
what if the stateIP doesn't match configIP but the PLs match?

Reply:
that handled by deleting ip that exist in state but not config for which there's a todo